### PR TITLE
Add benchmarks for melt migration

### PR DESCRIPTION
**Pull request by jdannberg**
Add a number of benchmarks together with documentation and scripts for how to run them, and a section in the manual. This includes the solitary wave benchmark and several test cases for linear and non-linear shear bands.

### CODE REVIEW

File: benchmark/shear_bands/README [hunk 1]
```diff
@@ -0,0 +1,38 @@
+~~~~~~ Instruction for the magmatic shear bands testcase for Newtonian rheology ~~~~~~
```
Comment (bangerth) [inline benchmark/shear_bands/README]: -> Instructions (plural)

File: benchmark/shear_bands/README [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+and 'plot_growth_over_angle.py' plots them and saves them as 'growth_rate_angle.pdf'. 
+
+The scripts 'run_plane_melt_bands.sh' and 'plot_convergence.py' can be used to
+generate data for the growth rate error of plane wave melt bands (compared to the rates
+predicted from linear stability analysis) and plot the error over the number of degrees 
+of freedom. 'run_plane_melt_bands.sh' outputs the data to the files 
+'plane_wave_melt_bands_8pi' and 'plane_wave_melt_bands_16pi', and 'plot_convergence.py' 
+plots them and saves the plot as 'growth_rate_error.pdf'. 
+
+~~~~~~ Instruction for the magmatic shear bands testcase for non-linear rheology ~~~~~~
```
Comment (bangerth) [inline benchmark/shear_bands/README]: -> Instructions (plural)

File: benchmark/shear_bands/colors.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#dark green
+#blue
+#pink
+#red
+#orange
+#light green
+c = ['#0A5F02', '#1c567a', '#814292', '#d7383b', '#fdae61', '#c0f8b8']
+c = ['#0A5F02', '#1c567a', '#814292', '#d7383b', '#fdae61', '#c0f8b8']
```
Comment (bangerth) [inline benchmark/shear_bands/colors.py]: I don't think you need both lines :-)

File: benchmark/shear_bands/notes.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+This directory contains the magmatic "shear bands" benchmark defined in Katz et al., 2006
```
Comment (bangerth) [inline benchmark/shear_bands/notes.txt]: period at the end

File: benchmark/shear_bands/shear_bands.cc [hunk 5]
```diff
@@ -0,0 +1,903 @@
...
+                                   "shear bands material",
+                                   "A material model that corresponds to the setup to"
+                                   "generate magmatic shear bands described in Katz et al., "
+                                   "Nature, 2006.")
+
+    ASPECT_REGISTER_POSTPROCESSOR(ShearBandsPostprocessor,
+                                  "shear bands statistics",
+                                  "A postprocessor that writes output for the porosity field, "
+                                  "which can be used to calculate the angle of the shear "
+                                  "bands in the model. ")
```
Comment (bangerth) [inline benchmark/shear_bands/shear_bands.cc]: no need for the space at the end here

File: benchmark/shear_bands/shear_bands.cc [hunk 6]
```diff
@@ -0,0 +1,903 @@
...
+
+    ASPECT_REGISTER_POSTPROCESSOR(ShearBandsGrowthRate,
+                                  "shear bands growth rate",
+                                  "A postprocessor computes the growth rate of melt bands as "
+                                  "presicted from linear stability analysis (Spiegelman, 2003), "
+                                  "and compares modelled and analytical solution. ")
```
Comment (bangerth) [inline benchmark/shear_bands/shear_bands.cc]: no need for the space at the end here

File: benchmark/solitary_wave/notes.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+This directory contains the "solitary wave" benchmark defined in Keller et al 2013
```
Comment (bangerth) [inline benchmark/solitary_wave/notes.txt]: -> et al., 2013.

File: benchmark/solitary_wave/solitary_wave.cc [hunk 8]
```diff
@@ -0,0 +1,1021 @@
...
+  if (indata[0]>inoutdata[0])
+    {
+      for (int i=0; i<*len; ++i)
+        inoutdata[i] = indata[i];
+    }
+}
+
+
+template <>
+void MPI_max_and_data(const double &local_max,
```
Comment (bangerth) [inline benchmark/solitary_wave/solitary_wave.cc]: if all you need is the specialization for `T=double`, why declare the general template above at all?

File: benchmark/solitary_wave/solitary_wave.cc [hunk 9]
```diff
@@ -0,0 +1,1021 @@
...
+       * for x --> +- infinity
+       *
+       * @param amplitude The amplitude of the solitary wave, which is always
+       * greater than 1.
+       */
+      double solitary_wave_solution (const double phi, const double amplitude)
+      {
+        AssertThrow(phi > 1.0 && phi <= amplitude,
+                    ExcMessage("The solitary wave solution can only be computed "
+                               "for porosities larger than the background porosity of 1"
```
Comment (bangerth) [inline benchmark/solitary_wave/solitary_wave.cc]: space missing at the end

File: doc/manual/manual.tex [hunk 10]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
 It would be instructive to compare the results of this benchmark using tracer particles with the 
 results using compositional fields.
 
+\subsubsection{The solitary wave benchmark}
+\label{sec:benchmark-solitary_wave}
+
+\textit{This section was contributed by Juliane Dannberg and is based on a section in \cite{DH2015} by Juliane Dannberg and Timo Heister.}
+
+One of the most widely used benchmarks for codes that model the migration of melt through a compacting and dilating matrix is the propagation of solitary waves ( \cite[e.g.]{SS11, KMK2013, Schm00}).
```
Comment (bangerth) [inline doc/manual/manual.tex]: -> see, e.g., \cite{...}

File: doc/manual/manual.tex [hunk 11]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
 It would be instructive to compare the results of this benchmark using tracer particles with the 
 results using compositional fields.
 
+\subsubsection{The solitary wave benchmark}
+\label{sec:benchmark-solitary_wave}
+
+\textit{This section was contributed by Juliane Dannberg and is based on a section in \cite{DH2015} by Juliane Dannberg and Timo Heister.}
+
+One of the most widely used benchmarks for codes that model the migration of melt through a compacting and dilating matrix is the propagation of solitary waves ( \cite[e.g.]{SS11, KMK2013, Schm00}). 
+The benchmark is intended to test the accuracy of the solution of the two--phase flow equations as described in Section \ref{sec:melt_transport} and makes use of the fact that there is an analytical solution for the shape of solitary waves that travel through a partially molten rock with a constant background porosity without changing their shape and with a constant wave speed.
```
Comment (bangerth) [inline doc/manual/manual.tex]: `two--phase` -> `two-phase`

File: doc/manual/manual.tex [hunk 12]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
 results using compositional fields.
 
+\subsubsection{The solitary wave benchmark}
+\label{sec:benchmark-solitary_wave}
+
+\textit{This section was contributed by Juliane Dannberg and is based on a section in \cite{DH2015} by Juliane Dannberg and Timo Heister.}
+
+One of the most widely used benchmarks for codes that model the migration of melt through a compacting and dilating matrix is the propagation of solitary waves ( \cite[e.g.]{SS11, KMK2013, Schm00}). 
+The benchmark is intended to test the accuracy of the solution of the two--phase flow equations as described in Section \ref{sec:melt_transport} and makes use of the fact that there is an analytical solution for the shape of solitary waves that travel through a partially molten rock with a constant background porosity without changing their shape and with a constant wave speed. 
+Here, we follow the setup of the benchmark as it is described in \cite{BR86}, which considers one-dimensional solitary waves:
```
Comment (bangerth) [inline doc/manual/manual.tex]: `:` -> `.`

File: doc/manual/manual.tex [hunk 13]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
+
+The model features a perturbation of higher porosity with the amplitude $A \phi_0$ in a uniform low-porosity ($\phi=\phi_0$) background.  Due to its lower density, melt migrates upwards, dilating the solid matrix at its front and compacting it at its end.
```
Comment (bangerth) [inline doc/manual/manual.tex]: with the amplitude -> with amplitude

File: doc/manual/manual.tex [hunk 14]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
+
+One of the most widely used benchmarks for codes that model the migration of melt through a compacting and dilating matrix is the propagation of solitary waves ( \cite[e.g.]{SS11, KMK2013, Schm00}). 
+The benchmark is intended to test the accuracy of the solution of the two--phase flow equations as described in Section \ref{sec:melt_transport} and makes use of the fact that there is an analytical solution for the shape of solitary waves that travel through a partially molten rock with a constant background porosity without changing their shape and with a constant wave speed. 
+Here, we follow the setup of the benchmark as it is described in \cite{BR86}, which considers one-dimensional solitary waves:
+
+The model features a perturbation of higher porosity with the amplitude $A \phi_0$ in a uniform low-porosity ($\phi=\phi_0$) background.  Due to its lower density, melt migrates upwards, dilating the solid matrix at its front and compacting it at its end. 
+
+Assuming constant shear and compaction viscosities and using a permeability law in form of 
+{%
```
Comment (bangerth) [inline doc/manual/manual.tex]: the braces are not wrong, but do they have a purpose?

Comment (bangerth) [inline doc/manual/manual.tex]: in form of -> in the form

Comment (bangerth) [inline doc/manual/manual.tex]: or "of the form"

File: doc/manual/manual.tex [hunk 15]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
+\begin{align*}
+k_\phi &= k_0 \phi^3, && \text{ implying a Darcy coefficient }
+K_D(\phi) = \frac{k_0}{\eta_f} \phi^3 , \\
```
Comment (bangerth) [inline doc/manual/manual.tex]: I suspect you want space after the `\text{...}`. Please double check with the pdf.

File: doc/manual/manual.tex [hunk 16]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
+\intertext{and the non-dimensionalization }
+x &= \delta x' 
+  && \text{ with the compaction length } \delta = \sqrt{K_D(\phi_0)(\xi + \frac{4}{3}\eta)} , \\
```
Comment (bangerth) [inline doc/manual/manual.tex]: probably same

File: doc/manual/manual.tex [hunk 17]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
+\phi &= \phi_0 \phi ' 
+  && \text{ with the background porosity } \phi_0 , \\
```
Comment (bangerth) [inline doc/manual/manual.tex]: and here, and elsewhere

File: doc/manual/manual.tex [hunk 18]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
+(\mathbf u_s, \mathbf u_f) &= u_0 (\mathbf u_s, \mathbf u_f)' 
+  && \text{ with the separation flux } \phi_0 u_0 = K_D(\phi_0) \Delta\rho g ,
+\end{align*}
+}
+%
+the analytical solution for the shape of the solitary wave can be written in the implicit form as:
```
Comment (bangerth) [inline doc/manual/manual.tex]: in the implicit form -> in implicit form

File: doc/manual/manual.tex [hunk 19]
```diff
@@ -7826,6 +7826,63 @@ \subsubsection{The Crameri et al.~benchmarks}
...
+
+\lstinputlisting[language=prmfile]{cookbooks/benchmarks/solitary_wave/solitary_wave.prm}
+
+The benchmark uses a custom model to generate the initial condition for the porosity field as specified by the analytical solution, and its own material model, which includes the additional material properties needed by models with melt migration, such as the permeability, melt density and compaction viscosity. The solitary wave postprocessor compares the porosity and pressure in the model to the analytical solution, and computes the errors for the shape of the porosity, shape of the compaction pressure and the phase speed. 
+We apply the negative phase speed of the solitary wave as a boundary condition for the solid velocity. This changes the reference frame, so that the solitary wave stays in the center of the domain, while the solid moves downwards. The temperature evolution does not play a role in this benchmark, so all temperature and heating-related parameters are disabled or set to zero. 
+
+And extensive discussion of the results and convergence behaviour can be found in \cite{DH2015}.
```
Comment (bangerth) [inline doc/manual/manual.tex]: -> behavior

**Comment by bangerth** [comment]
Comments:

1/ I haven't even attempted to really read through the code. I assume that it's correct.
2/ Are the formulas for melt migration ever specified anywhere in the manual? I don't recall. If not, it would be nice to add a section somewhere.
3/ Is it possible to add a simplified version of the benchmark as a testcase?

**Comment by jdannberg** [comment]
Okay, I addressed your commits. To reply to your comments:
1/ I tested all of the benchmarks again before submitting the pull request, so I will also just assume that the code is correct :-)
2/ The equations were already part of the first melt pull request (that we already merged). 
3/ Same goes for the tests. 
Thanks for reviewing!

**Comment by bangerth** [comment]
OK. Please address the issue about `max_and_data` and then merge.

**Comment by jdannberg** [comment]
We still have an issue with the stabilization for one of the benchmarks in this pull request, so please don't merge for now.

**Comment by jdannberg** [comment]
Okay, I have updated the one file that had the issue with the stabilization (it works now with the new stabilization parameter), so this should be good to go (unless you have any other comments).

**Pull request merged**